src/library_manager/lib_manager.c

```diff
@@ -69,6 +69,68 @@ static int lib_manager_load_data_from_storage(void __sparse_cache *vma, void *s_
 	return 0;
 }
 
+static int lib_manager_load_module(uint32_t module_id, struct sof_man_module *mod,
+				   struct sof_man_fw_desc *desc);
+static int lib_manager_unload_module(uint32_t module_id, struct sof_man_module *mod,
+				     struct sof_man_fw_desc *desc);
+
+static int lib_manager_load_aux(struct sof_man_module *mod,
+				struct sof_man_fw_desc *desc, uint32_t lib_id)
+{
+	struct ext_library *ext_lib = ext_lib_get();
+
+	/* There are modules marked as lib_code. This is code shared between several modules inside
+	 * the library. Load all lib_code modules with first none lib_code module load.
+	 */
+	if (!mod->type.lib_code)
+		ext_lib->mods_exec_load_cnt++;
+
+	if (ext_lib->mods_exec_load_cnt == 1) {
+		struct sof_man_module *module_entry =
+		    (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(0));
+
+		for (size_t idx = 0; idx < desc->header.num_module_entries;
+				++idx, ++module_entry) {
+			if (module_entry->type.lib_code) {
+				int ret = lib_manager_load_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
+								  idx, mod, desc);
+				if (ret < 0)
+					return ret;
+			}
+		}
+	}
+
+	return 0;
+}
+
+static void lib_manager_unload_aux(struct sof_man_module *mod,
+				   struct sof_man_fw_desc *desc, uint32_t lib_id)
+{
+	struct ext_library *ext_lib = ext_lib_get();
+
+	/* There are modules marked as lib_code. This is code shared between several modules inside
+	 * the library. Unload all lib_code modules with last none lib_code module unload.
```

Comment:
@ujfalusi didn't we make the assumption that a library contains ONE module?

IIRC the firmwares are identified with a UUID, not sure how the library is identified in Linux

Reply:
@plbossart in an earlier version I removed this whole "lib_code" support because, well, it's unused so far and it's unclear if and when and how it would be used. I got complains. So, I keep it as is, just move out to functions

Reply:
No, the kernel side does not assume anything, it is prepared for the module bundles. The kernel tracks the library itself and then checks the module(s) inside:
b700672e2250 ("ASoC: SOF: Intel/IPC4: Support for external firmware libraries")

The lib_code does not have UUID, but in case a library have multiple modules then there should be symlinks:
```
module_bundle.bin
<UUID1>.bin -> module_bundle.bin
<UUID2>.bin -> module_bundle.bin
<UUID3>.bin -> module_bundle.bin
```
The kernel will load a library for a specific module UUID, so multiple UUID pointing to the same library will load that library with all of it's modules once.